distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
-            (bins[i][0]*bins[i][1] + bins[i+1][0]*bins[i+1][1])
-            / (bins[i][1] + bins[i+1][1]),
-            bins[i][1] + bins[i+1][1]
-        )
-        del bins[i+1]
+            diffs = [
+                (i - 1, _weighted_diff(h, b, h.bins[i - 1]))
+                for i, b in enumerate(h.bins[1:], start=1)
+            ]
+            i, _ = min(diffs, key=itemgetter(1))
+
+        v1, f1 = h.bins[i]
+        v2, f2 = h.bins[i + 1]
+        h.bins[i] = (v1 * f1 + v2 * f2) / (f1 + f2), f1 + f2
+        del h.bins[i + 1]
+
         if h.diffs is not None:
             del h.diffs[i]
             _update_diffs(h, i)
             h.min_diff = min(h.diffs)
 
-    h.bins = bins
     return h
 
 
-def _trim_in_place(h, value, count, i):
-    bins = h.bins
-
-    bins[i] = (
-        (bins[i][0]*bins[i][1] + value*count)
-        / (bins[i][1] + count),
-        bins[i][1] + count
-    )
-
-    h.bins = bins
+def _trim_in_place(h: Distogram, value: float, c: int, i: int):
+    v, f = h.bins[i]
+    h.bins[i] = (v * f + value * c) / (f + c), f + c
     _update_diffs(h, i)
     return h
 
 
-def _compute_diffs(h):
-    diffs = []
-    bins = h.bins
-    h.min_diff = None
-    for index in range(1, len(bins)):
-        diff = _weighted_diff(h, h.bins[index], h.bins[index-1])
-        diffs.append(diff)
-        if h.min_diff is None or diff < h.min_diff:
-            h.min_diff = diff
+def _compute_diffs(h: Distogram) -> List[float]:
+    diffs = [
+        _weighted_diff(h, b2, b1)
+        for b1, b2 in zip(h.bins[:-1], h.bins[1:])
+    ]
+    h.min_diff = min(diffs)
 
     return diffs
 
 
-def _linear_index(bins, value):
-    ratio = float((value - bins[0][0]) / (bins[-1][0] - bins[0][0]))
-    index = round(ratio * len(bins))
-    if index > 0 and index < (len(bins) - 1):
-        if value > bins[index][0] and value < bins[index+1][0]:
-            return index+1
-        if value > bins[index-1][0] and value < bins[index][0]:
-            return index
-
-    return 0
-
-
-def _search_in_place_index(h, new_value, index):
-    bins = h.bins
-
+def _search_in_place_index(h: Distogram, new_value: float, index: int) -> int:
     if h.diffs is None:
         h.diffs = _compute_diffs(h)
 
-    min_diff = h.min_diff
-    i_bin = None
-
     if index > 0:
-        diff1 = _weighted_diff(h, (new_value, 1), bins[index-1])
-        diff2 = _weighted_diff(h, bins[index], (new_value, 1))
-        if diff1 < diff2:
-            diff = diff1
-            i_bin = index-1
-        else:
-            diff = diff2
-            i_bin = index
+        diff1 = _weighted_diff(h, (new_value, 1), h.bins[index - 1])
+        diff2 = _weighted_diff(h, h.bins[index], (new_value, 1))
+
+        i_bin, diff = (index - 1, diff1) if diff1 < diff2 else (index, diff2)
+
+        return i_bin if diff < h.min_diff else -1
 
-        if diff < min_diff:
-            return i_bin
-        else:
-            return -1
     return -1
 
 
-def update(h, value, count=1):
```

Comment:
There were a few places where an argument name shadowed the name of a function from out of scope. This is generally bad practice in Python so I shortened the name. We could easily go back on this if you want.

Reply:
It shadows distogram.count. I have renamed it. (I'll make the push after addressing all comments)